### PR TITLE
fix: Parse Server option `extendSessionOnUse` not working for session lengths < 24 hours

### LINKED ISSUE #8981: extendSessionOnUse not working as expected
### New Issue Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).
- [x] I can reproduce the issue with the [latest version of Parse Server](https://github.com/parse-community/parse-server/releases). 

### Issue Description
Issue with 'extendSessionOnUse'
 I don't understand how this feature works.
So this is where the magic happens and the session gets extended.

```
 const lastUpdated = new Date((_session = session) === null || _session === void 0 ? void 0 : _session.updatedAt);
      const yesterday = new Date();
      yesterday.setDate(yesterday.getDate() - 1);
      if (lastUpdated > yesterday || !session) {
        return;
      }
      const expiresAt = config.generateSessionExpiresAt();
      await new _RestWrite.default(config, master(config), '_Session', {
        objectId: session.objectId
      }, {
        expiresAt: Parse._encode(expiresAt)
      }).execute();
```

But looking at this piece of code it takes lastUpdated and yesterday and if lastUpdated is greater than yesterday it returns.

So if I've doing 60 minute sessions, well this will never actually extend the session. So for this to work a session needs to be at least 24 hours.

This isn't documented anywhere and it's counterintuitive.

Am I missing something.

### Steps to reproduce
Enable extendSessionOnUse.

Set session to 60 minutes.

Perform actions for authenticated user and look at session expiry. it doesn't update.

Now set session to 48 hours. 

Perform actions and notice that session expiry does update.

### Actual Outcome

I expect Session expiry to update for any session duration.

### Expected Outcome

I expect Session expiry to update for any session duration.
### Environment

NodeJS 21

Server
- Parse Server version: `6.4
- Operating system: `Ubuntu 22`
- Local or remote host (AWS, Azure, Google Cloud, Heroku, Digital Ocean, etc): `local`

Database
- System (MongoDB or Postgres): `MongoDB
- Database version: `6.0`
- Local or remote host (MongoDB Atlas, mLab, AWS, Azure, Google Cloud, etc): `local`

Client
- SDK (iOS, Android, JavaScript, PHP, Unity, etc): `REST`
- SDK version: `REST`

### Logs

N/A

### PR DESCRIPTION
**Pull request by vivekjoshi556**
## Pull Request
This PR helps with dynamically extending session based on session Length.

## Issue

Closes: #8981 

## Approach
I followed a similar approach to what @mman suggested [here](https://github.com/parse-community/parse-server/issues/8981#issuecomment-1976147677).

## Tasks
- [x] Add tests
- [ ] Add changes to documentation (guides, repository pages, code comments)

### CODE REVIEW

**Comment by vivekjoshi556** [comment]
Hello @mman,
You mentioned that you would beta test. Please look at the solution and let me know what you think.

**Comment by mman** [comment]
Thanks a lot @vivekjoshi556, I will take a look first thing next week, now OOO.

**Comment by vivekjoshi556** [comment]
Hey @mman. Just wanted to check if you got some time to look at the PR.

**Comment by mman** [comment]
@vivekjoshi556 The code looks nice and clean, the tests as well. Will try to deploy to my stage env later today.

**Comment by vivekjoshi556** [comment]
Hey @mman Just wanted to check in again.

@mtrezza if you have inputs as well, please let me know. 

There is a minor issue with linting I'll fix it in the next push, anything else let me know.

File: src/Auth.js [hunk 1]
```diff
@@ -67,6 +67,23 @@ function nobody(config) {
   return new Auth({ config, isMaster: false });
 }
 
+// A helper to check whether session should be updated based on last update time & session length.
+function shouldUpdateSessionExpiry(config, session) {
```
Comment (mtrezza) [inline src/Auth.js]:
```suggestion
/**
 * Checks whether session should be updated based on last update time & session length.
 */
function shouldUpdateSessionExpiry(config, session) {
```

File: spec/Auth.spec.js [hunk 2]
```diff
@@ -260,3 +260,58 @@ describe('Auth', () => {
...
+      sessionLength: 120,
+      sessionUpdatedAt: 50,
+      result: false,
+    },
+  ];
+
+  tests.forEach(({ title, sessionLength, sessionUpdatedAt, result }) => {
+    it(`shouldUpdateSessionExpiry() when ${title}`, async () => {
+      const { shouldUpdateSessionExpiry } = require('../lib/Auth');
+      let update = new Date();
```
Comment (mtrezza) [inline spec/Auth.spec.js]:
```suggestion
      const update = new Date();
```

File: src/Auth.js [hunk 3]
```diff
@@ -67,6 +67,25 @@ function nobody(config) {
...
+    resetAfter = 3600; // 1 hour
+  } else {
+    resetAfter = 60; // Every minute
+  }
+
+  const lastUpdated = new Date(session?.updatedAt);
+  const skipRange = new Date();
+  skipRange.setTime(skipRange.getTime() - resetAfter * 1000);
+  return lastUpdated <= skipRange;
+}
```
Comment (mtrezza) [inline src/Auth.js]: Could you explain this logic? 

Edit: got it, looking at https://github.com/parse-community/parse-server/issues/8981#issuecomment-1976147677

**Comment by mtrezza** [comment]
This fix just shifts the problem downwards from 24 hours to 1 minute. Because for a session length of <1 minute, the problem in https://github.com/parse-community/parse-server/issues/8981 persists.

I wonder whether we need such a complex logic for this. If I understand https://github.com/parse-community/parse-server/issues/8981#issuecomment-1976147677 correctly, the purpose of this logic is to find a middle ground between:
- extending the session on every request --> con: high DB writes
- ensuring that the session extends before it expires --> con: session would never get extended

Wouldn't it be simpler to extend the session when the remaining session token validity time reached < 50% of the session length? 

For example:
- session length 1 year --> session extends if remaining time is < 182 days
- session length 1 day --> session extends if remaining time is < 12 hours
- session length 1 hour --> session extends if remaining time is < 30 mins
- session length 1 minute --> session extends if remaining time is < 30 secs

Pros:
- Simple to document and understand
- This adapts dynamically to the session length, instead of some arbitrary extension trigger times
- It's easy to make this configurable via a new Parse Server option in the future, by letting a developer configure the threshold, e.g. 50% as in examples above.

And could you please document this new behavior in the Parse Server option `extendSessionOnUse`?

File: spec/Auth.spec.js [hunk 4]
```diff
@@ -260,3 +260,58 @@ describe('Auth', () => {
...
+      const { shouldUpdateSessionExpiry } = require('../lib/Auth');
+      const update = new Date();
+      update.setTime(update.getTime() - sessionUpdatedAt * 1000);
+      const res = shouldUpdateSessionExpiry(
+        { sessionLength: sessionLength },
+        { updatedAt: update }
+      );
+
+      expect(res).toBe(result);
+    });
```
Comment (mtrezza) [inline spec/Auth.spec.js]: Please refactor by moving the loop into the test. We don't need a separate test log for every param. Considering https://github.com/parse-community/parse-server/pull/9113#issuecomment-2104597113, 1 simple test would be enough, not sure whether you need a loop at all here.

**Review by mtrezza — CHANGES_REQUESTED** [review]

**Comment by mtrezza** [comment]
@vivekjoshi556 Just a friendly ping here; if you could simplify the logic and add the functionality description to the option docs (this repo, not the docs repo), then we could go ahead and merge.

**Comment by vivekjoshi556** [comment]
So sorry just got a bit swamped with work. Give me some time

**Comment by vivekjoshi556** [comment]
@mtrezza I have made the mentioned changes. Please let me know if anything else needs to be done.

File: src/Options/index.js [hunk 5]
```diff
@@ -228,7 +228,7 @@ export interface ParseServerOptions {
   /* Session duration, in seconds, defaults to 1 year
   :DEFAULT: 31536000 */
   sessionLength: ?number;
-  /* Whether Parse Server should automatically extend a valid session by the sessionLength
+  /* Whether Parse Server should automatically extend a valid session by the sessionLength. The session will be extended when a request is received after half of the session's lifetime has passed.
```
Comment (mtrezza) [inline src/Options/index.js]:
```suggestion
  /* Whether Parse Server should automatically extend a valid session by the sessionLength. In order to reduce the number of session updates in the database, a session will only be extended when a request is received after at least half of the current session's lifetime has passed.
```

File: src/Options/docs.js [hunk 6]
```diff
@@ -47,7 +47,7 @@
  * @property {String} encryptionKey Key for encrypting your files
  * @property {Boolean} enforcePrivateUsers Set to true if new users should be created without public read and write access.
  * @property {Boolean} expireInactiveSessions Sets whether we should expire the inactive sessions, defaults to true. If false, all new sessions are created with no expiration date.
- * @property {Boolean} extendSessionOnUse Whether Parse Server should automatically extend a valid session by the sessionLength
+ * @property {Boolean} extendSessionOnUse Whether Parse Server should automatically extend a valid session by the sessionLength. The session will be extended when a request is received after half of the session's lifetime has passed.
```
Comment (mtrezza) [inline src/Options/docs.js]:
```suggestion
 * @property {Boolean} extendSessionOnUse Whether Parse Server should automatically extend a valid session by the sessionLength. In order to reduce the number of session updates in the database, a session will only be extended when a request is received after at least half of the current session's lifetime has passed.
```

File: src/Options/Definitions.js [hunk 7]
```diff
@@ -259,7 +259,8 @@ module.exports.ParseServerOptions = {
   },
   extendSessionOnUse: {
     env: 'PARSE_SERVER_EXTEND_SESSION_ON_USE',
-    help: 'Whether Parse Server should automatically extend a valid session by the sessionLength',
+    help:
+      "Whether Parse Server should automatically extend a valid session by the sessionLength. The session will be extended when a request is received after half of the session's lifetime has passed.",
```
Comment (mtrezza) [inline src/Options/Definitions.js]:
```suggestion
    help:
      "Whether Parse Server should automatically extend a valid session by the sessionLength. In order to reduce the number of session updates in the database, a session will only be extended when a request is received after at least half of the current session's lifetime has passed.",
```

**Review by mtrezza — APPROVED** [review]

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [7.1.0-alpha.9](https://github.com/parse-community/parse-server/releases/tag/7.1.0-alpha.9)

**Comment by mman** [comment]
@vivekjoshi556 Excellent, very clean code, putting into production :)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [7.1.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/7.1.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [7.1.0](https://github.com/parse-community/parse-server/releases/tag/7.1.0)

**Pull request merged**